File: javascript/id3fix.js
const os = require('os');
const fs = require('fs');
const requests = require('requests');

const MP3File = require('mp3_tagger').MP3File;
const { VERSION_1, VERSION_2, VERSION_BOTH } = require('mp3_tagger');

function checkartist(all, artists) {
    let result = 0;
    for (let artist of all) {
        if (artists.includes(artist.name)) {
            result += 1;
        }
    }
    return result === artists.length;
}

function initquery(filter) {
    const qstring = `http://localhost:3000/cloudsearch?keywords=${filter.replace('')}`;
    const x = requests.get(qstring);
    const result = JSON.parse(x.text).result;
    if (result === null || result === {} || result.songCount === 0) {
        console.log(`can not find song from ${filter}`);
        return null;
    }
    return result;
}

function findtrack(result, title, artists, albumname) {
    if (result === null) {
        return 0;
    }
    let track = 0;
    for (let song of result.songs) {
        if (song.name !== title) {
            continue;
        }
        if (!checkartist(song.ar, artists)) {
            continue;
        }
        const album = song.al;
        if (album.name !== albumname) {
            console.log(`album name (${album.name}) not match with ${albumname}`);
        }
        const x = requests.get(`http://localhost:3000/album?id=${album.id}`);
        const songs = JSON.parse(x.text).songs;
        for (let index = 0; index < songs.length; index++) {
            if (songs[index].name === title) {
                track = index + 1;
                break;
            }
        }
        if (track > 0) {
            return track;
        }
    }
    return track;
}

function savetrack(mp3, track) {
    mp3.track = String(track);
    mp3.save();
    return track;
}

function addtrack(mp3file) {
    console.log(`processing on .. ${mp3file}`);
    const mp3 = new MP3File(mp3file);
    mp3.set_version(VERSION_2);
    const tags = mp3.get_tags();
    if (tags.song === null) {
        console.log(`${mp3file} is invalid mp3, or missing key tags`);
        return -1;
    }
    if (tags.track !== null) {
        return -1;
    }
    const song = tags.song.replace('\x00', '');
    const album = tags.album.replace('\x00', '');
    const artists = tags.artist.replace('\x00', '').split('、');
    console.log(`fix track ... ${mp3file} with tags ${tags}`);
    let result = initquery(artists.join('') + ' ' + song);
    let track = findtrack(result, song, artists, album);
    if (track !== 0) {
        return savetrack(mp3, track);
    }
    result = initquery(song);
    track = findtrack(result, song, artists, album);
    if (track !== 0) {
        return savetrack(mp3, track);
    }
    result = initquery(artists.join(''));
    track = findtrack(result, song, artists, album);
    if (track !== 0) {
        return savetrack(mp3, track);
    }
    return track;
}

const appendix = "mp3";
const root = '/Users/shichang/Downloads/temp/石头/';
const candidates = [];
candidates.push(root);
let index = 0;
while (index < candidates.length) {
    const path = candidates[index];
    const dirs = fs.readdirSync(path);
    console.log(`process ${path}`);
    for (let file of dirs) {
        const cur = path + file;
        if (fs.statSync(cur).isDirectory()) {
            candidates.push(cur + '/');
            console.log(`add ${cur}`);
        }
        if (file.endsWith(appendix)) {
            if (addtrack(cur) === 0) {
                console.log(`no track info find for ${file}`);
            }
        }
    }
    index += 1;
}
console.log(candidates);

File: python/utils/id3fix.py
import os, sys, requests, json
from mp3_tagger import MP3File, VERSION_1, VERSION_2, VERSION_BOTH

def checkartist(all, artists):
    result = 0
    for artist in all:
        if artist.get('name') in artists:
            result += 1
    return result == len(artists)

def initquery(filter):
    qstring = 'http://localhost:3000/cloudsearch?keywords=%s' % filter.replace('#','')[0:50]
    x = requests.get(qstring)
    result = json.loads(x.text).get('result')
    if (result is None) or (result == {}) or (result.get('songCount') == 0):
        print('can not find song from %s' % filter)
        return None
    return result

def findtrack(result, title, artists, albumname):
    if result == None:
        return 0
    track = 0
    for song in result.get('songs'):
        if (song.get('name') != title):
            continue
        if not checkartist(song.get('ar'), artists):
            continue
        album = song.get('al')
        if album.get('name') != albumname:
            print('album name (%s) not match with %s' % (album.get('name'), albumname))

        x = requests.get('http://localhost:3000/album?id=%d' % album.get('id'))
        songs = json.loads(x.text).get('songs')
        for index in range(len(songs)):
            if songs[index].get('name') == title:
                track = index + 1
                break
        if (track > 0):
            return track
    return track

def savetrack(mp3, track):
#    print('save track as %d for %s' % (track, mp3.get_tags()))
    mp3.track = str(track)
    mp3.save()
    return track

def addtrack(mp3file):
    print('processing on .. %s' % mp3file)
    mp3 = MP3File(mp3file)
    mp3.set_version(VERSION_2)
    tags = mp3.get_tags()
    if (tags.get('song') is None):
        print('%s is invalid mp3, or missing key tags' % mp3file)
        return -1
    if (tags.get('track') is not None):
        return -1
    song = tags.get('song').replace('\x00','')
    album = tags.get('album').replace('\x00','')
    artists = tags.get('artist').replace('\x00','').split('、')
    
    print("fix track ... %s with tags %s", (mp3file, tags))
    result = initquery(' '.join(artists) + '' + ' ' + song)
    track = findtrack(result, song, artists, album)
    if (track != 0):
        return savetrack(mp3, track)

    result = initquery(song)
    track = findtrack(result, song, artists, album)
    if (track != 0):
        return savetrack(mp3, track)

    result = initquery(' '.join(artists))
    track = findtrack(result, song, artists, album)
    if (track != 0):
        return savetrack(mp3, track)
    return track

#addtrack('/Users/shichang/Music/take/take/月夏寻星计划/Girl - NuclearBomb牛客帮.mp3')


appendix = "mp3"
root = '/Users/shichang/Downloads/temp/FromMac/'
candidates = []
candidates.append(root)
index = 0
while (index < len(candidates)):
    path = candidates[index]
    dirs = os.listdir(path)
    print('process %s' % path)
    for file in dirs:
        cur = path + file
        if (os.path.isdir(cur)):
            candidates.append(cur + '/')
            print('add %s' % cur)
        if (file.endswith(appendix)):
            if (addtrack(cur) == 0):
                print('no track info find for %s' % file)
    index += 1
    
print(candidates)
